src/python.cc

```diff
@@ -956,32 +950,147 @@ ModelInstanceState::SetupChildProcess()
+      (char**)&parent_mutex, sizeof(pthread_mutex_t), parent_mutex_offset));
+  CreateIPCMutex(&parent_mutex);
+
+  pthread_cond_t* parent_cv;
+  off_t parent_cv_offset;
+  RETURN_IF_EXCEPTION(shm_pool_->Map(
+      (char**)&parent_cv, sizeof(pthread_cond_t), parent_cv_offset));
+  CreateIPCCondVariable(&parent_cv);
+
+  parent_cond_ = parent_cv;
+  parent_mutex_ = parent_mutex;
+
+  off_t ipc_offset;
+  RETURN_IF_EXCEPTION(
+      shm_pool_->Map((char**)&ipc_message_, sizeof(IPCMessage), ipc_offset));
+
+  uint64_t model_version = model_state->Version();
+  const char* model_path = model_state->RepositoryPath().c_str();
+
+  std::stringstream ss;
+  // Use <path>/version/model.py as the model location
+  ss << model_path << "/" << model_version << "/model.py";
+  model_path_ = ss.str();
+  struct stat buffer;
+
+  // Check if model.py exists
+  if (stat(model_path_.c_str(), &buffer) != 0) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("model.py does not exist in the model repository path: " + model_path_)
+            .c_str());
+  }
+
+  // Path to the extracted Python env
+  std::string python_execution_env = "";
+
+  if (model_state->PythonExecutionEnv() != "") {
+    try {
+      python_execution_env =
+          model_state->StateForBackend()->env_manager->ExtractIfNotExtracted(
+              model_state->PythonExecutionEnv());
+    }
+    catch (PythonBackendException& pb_exception) {
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INTERNAL, pb_exception.what());
+    }
+    path_to_activate_ = python_execution_env + "/bin/activate";
+    path_to_libpython_ = python_execution_env + "/lib";
+    if (python_execution_env.length() > 0 && !FileExists(path_to_activate_)) {
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INTERNAL,
+          (std::string("Path ") + path_to_activate_ +
+           " does not exist. The Python environment should contain an "
+           "'activate' script.")
+              .c_str());
+    }
+  }
+
+  parent_pid_ = getpid();
+  pthread_mutex_lock(parent_mutex_);
+  RETURN_IF_ERROR(StartStubProcess());
+
   return nullptr;
 }
 
 ModelInstanceState::~ModelInstanceState()
 {
   if (initialized_) {
-    // Create Python inference requests
-    RequestBatch* request_batch;
-    off_t request_batch_offset;
-    shm_pool_->Map(
-        (char**)&request_batch, sizeof(RequestBatch), request_batch_offset);
-    //"failed to create request batch in shared memory.");
-    request_batch->batch_size = 0;
-    ipc_message_->request_batch = request_batch_offset;
-    NotifyChild();
-
-    // Wait for child notification
-    pthread_cond_wait(parent_cond_, parent_mutex_);
-    pthread_mutex_unlock(parent_mutex_);
+    ipc_message_->health = false;
+
+    // Sleep 1 second so that the child process has a chance to change the
+    // health variable
+    sleep(1);
+
+    if (ipc_message_->health) {
+      // Create Python inference request
```

Comment:
Why are you sending a request to the stub at this point? Is this how you tell the stub to exit? Why not a more direct way... at the very least this comment needs updating.

Reply:
It is sending a request to trigger the shutdown and calling of `finalize` on the Python stub side. I agree that the messaging mechanism needs an update/refactor. When the main process creates a request of size zero, it triggers the shutdown on the stub side. I'll update the comment to be more clear.